cf-custom-resources/lib/bucket-cleaner.js

```diff
@@ -0,0 +1,200 @@
+ * @param {string} [reason] reason for failure, if any, to convey to the user
+ * @returns {Promise} Promise that is resolved on success, or rejected on connection error or HTTP error response
+ */
+let report = function (
+  event,
+  context,
+  responseStatus,
+  physicalResourceId,
+  responseData,
+  reason
+) {
+  return new Promise((resolve, reject) => {
+    const https = require("https");
+    const { URL } = require("url");
+
+    var responseBody = JSON.stringify({
+      Status: responseStatus,
+      Reason: reason,
+      PhysicalResourceId: physicalResourceId || context.logStreamName,
+      StackId: event.StackId,
+      RequestId: event.RequestId,
+      LogicalResourceId: event.LogicalResourceId,
+      Data: responseData,
+    });
+
+    const parsedUrl = new URL(event.ResponseURL || defaultResponseURL);
+    const options = {
+      hostname: parsedUrl.hostname,
+      port: 443,
+      path: parsedUrl.pathname + parsedUrl.search,
+      method: "PUT",
+      headers: {
+        "Content-Type": "",
+        "Content-Length": responseBody.length,
+      },
+    };
+
+    https
+      .request(options)
+      .on("error", reject)
+      .on("response", (res) => {
+        res.resume();
+        if (res.statusCode >= 400) {
+          reject(new Error(`Error ${res.statusCode}: ${res.statusMessage}`));
+        } else {
+          resolve();
+        }
+      })
+      .end(responseBody, "utf8");
+  });
+};
+
+/**
+ * Delete all objects in a bucket.
+ *
+ * @param {string} bucketName Name of the bucket to be cleaned.
+ */
+const cleanBucket = async function (bucketName) {
+  let s3 = new aws.S3();
+  // Make sure the bucket exists.
+  try {
+    await s3.headBucket({ Bucket: bucketName }).promise();
+  } catch (err) {
+    if (err.name === "ResourceNotFoundException") {
+      return;
+    }
+    throw err;
+  }
+  const listObjectVersionsParam = {
+    Bucket: bucketName
+  }
+  while (true) {
+    const listResp = await s3.listObjectVersions(listObjectVersionsParam).promise();
+    let objectsToDelete = []
+    for (const object of listResp.Versions) {
+      objectsToDelete.push({
+        Key: object.Key,
+        VersionId: object.VersionId
+      })
+    }
+    // After deleting other versions, remove delete markers version.
+    // For info on "delete marker": https://docs.aws.amazon.com/AmazonS3/latest/dev/DeleteMarker.html
+    for (const deleteMarker of listResp.DeleteMarkers) {
+      objectsToDelete.push({
+        Key: deleteMarker.Key,
+        VersionId: deleteMarker.VersionId
+      })
+    }
+    if (objectsToDelete.length === 0) {
+      return
+    }
+    const delResp = await s3.deleteObjects({
+      Bucket: bucketName,
+      Delete: {
+        Objects: objectsToDelete,
+        Quiet: true
+      }
+    }).promise()
+    if (delResp.Errors.length > 0) {
+      throw new AggregateError([new Error(`${delResp.Errors.length}/${objectsToDelete.length} objects failed to delete`),
```

Comment:
> I have a question on AggregateError. Isn't this already an inbuilt Error in js 16 🤔

That's what I thought too but it doesn't seem to be the case.

> And also do we need to log all the error messages instead of only first error or is the first error good.

I just migrate what we do currently and I think the first error msg can be a good start for triaging.